typed-protocols/test/Test/Network/TypedProtocol/PingPong.hs

```diff
@@ -0,0 +1,172 @@
+import           Data.Functor.Identity (Identity (..))
+import           Data.List (foldl')
+
+
+import           Network.TypedProtocol.Codec (transformCodec)
+import           Network.TypedProtocol.Channel
+  ( Channel
+  , createConnectedChannels
+  , createPipeConnectedChannels
+  )
+import           Network.TypedProtocol.Driver (runPeer)
+import           Network.TypedProtocol.Proofs (connect, connectPipelined)
+
+import           Network.TypedProtocol.PingPong.Type (pingPongAgencyProofs)
+import           Network.TypedProtocol.PingPong.Client
+  ( pingPongClientPeer
+  , pingPongClientPeerSender
+  )
+import           Network.TypedProtocol.PingPong.Codec
+  ( codecPingPong
+  , codecPingPong
+  )
+import           Network.TypedProtocol.PingPong.Server
+  ( pingPongServerPeer
+  )
+import           Network.TypedProtocol.PingPong.Examples
+  ( pingPongClientCount
+  , pingPongSenderCount
+  , pingPongServerCount
+  )
+import           Network.TypedProtocol.PingPong.Direct (direct)
+
+import           Test.Tasty
+import           Test.QuickCheck
+import           Test.Tasty.QuickCheck (testProperty)
+
+tests :: TestTree
+tests = testGroup "Network.TypedProtocol.PingPong"
+  [ testProperty "direct" prop_direct
+  , testProperty "connect" prop_connect
+  , testProperty "connect_piplined ST" prop_connect_pipelined_ST
+  , testProperty "connect_piplined IO" prop_connect_pipelined_IO
+  , testProperty "channel ST" prop_channel_ST
+  , testProperty "channel IO" prop_channel_IO
+  , testProperty "pipe" prop_pipe
+  ]
+
+runExperiment
+  :: forall m n.
+     ( MonadSTM m
+     , MonadTimer m
+     , MonadProbe m
+     , MonadRunProbe m n
+     )
+  => (Probe m Property -> m ())
+  -> n Property
+runExperiment exp_ = isValid <$> withProbe exp_
+ where
+  isValid = foldl' (\acu (_,p) -> acu .&&. p) (property True)
+
+prop_direct
+  :: Positive Int
+  -> Property
+prop_direct (Positive x) =
+  let c = fromIntegral x
+  in case runIdentity $ direct (pingPongClientCount c) pingPongServerCount of
+    (_, c') -> c === c'
+
+prop_connect
+  :: Positive Int
+  -> Property
+prop_connect (Positive x) =
+  let c = fromIntegral x
+  in case runIdentity $ connect pingPongAgencyProofs (pingPongClientPeer $ pingPongClientCount c) (pingPongServerPeer pingPongServerCount) of
+    (_, c') -> c === c'
+
+connect_pipelined_experiment
+  :: ( MonadSTM m
+     , MonadProbe m
+     )
+  => Positive Int
+  -> Probe m Property
+  -> m ()
+connect_pipelined_experiment (Positive x) probe = do
+  var <- atomically $ newTVar 0
+  let c = fromIntegral x
+      client = pingPongSenderCount var c
+  (_, b) <- connectPipelined pingPongAgencyProofs (pingPongClientPeerSender client) (pingPongServerPeer pingPongServerCount)
+  res <- atomically $ readTVar var
+  probeOutput probe (c === b)
+  probeOutput probe (c === res)
+
+prop_connect_pipelined_ST
+  :: Positive Int
+  -> Property
+prop_connect_pipelined_ST p = runST $ runExperiment @(Free (SimF _))
```

Comment:
In `3822e3a73525b8e744936c4fea894bb98493d245` in branch `coot/block-fetch-v2` I defined the `runPropIO = ioProperty . runExepriment` and `runPropSimM = runST . runExperiment`.  Then one does not need to use `TypeApplication` extension.  Either cherry-pick it, or make an issue, and I'll do that later.

Reply:
:+1:

Reply:
It also reduces the amount of functions one needs to define, since you can just "inline" them in the list of tests.